Sources/GRPC/GRPCClientStateMachine.swift

```diff
@@ -0,0 +1,588 @@
+    headers.add(contentsOf: options.customMetadata)
+
+    // Add a tracing header if the user specified it.
+    if let headerName = options.requestIDHeader {
+      headers.add(name: headerName, value: requestID)
+    }
+
+    return HTTPRequestHead(
+      version: HTTPVersion(major: 2, minor: 0),
+      method: options.cacheable ? .GET : .POST,
+      uri: path,
+      headers: headers
+    )
+  }
+
+  /// Parses the response headers ("Response-Headers" in the specification) from server into
+  /// a `ReadState`.
+  ///
+  /// See: https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#responses
+  ///
+  /// - Parameter headers: The headers to parse.
+  private func parseResponseHeaders(
+    _ head: HTTPResponseHead,
+    responseArity: MessageCount,
+    logger: Logger
+  ) -> Result<ReadState, RecieveResponseHeadError> {
+    // From: https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#responses
+    //
+    // "Implementations should expect broken deployments to send non-200 HTTP status codes in
+    // responses as well as a variety of non-GRPC content-types and to omit Status & Status-Message.
+    // Implementations must synthesize a Status & Status-Message to propagate to the application
+    // layer when this occurs."
+    guard head.status == .ok else {
+      return .failure(.invalidHTTPStatus(head.status))
+    }
+
+    guard head.headers["content-type"].first?.starts(with: "application/grpc") ?? false else {
+      return .failure(.invalidContentType)
+    }
+
+    // What compression mechanism is the server using, if any?
+    let compression = CompressionMechanism(value: head.headers[GRPCHeaderName.encoding].first)
+
+    // From: https://github.com/grpc/grpc/blob/master/doc/compression.md
+    //
+    // "If a server sent data which is compressed by an algorithm that is not supported by the
+    // client, an INTERNAL error status will occur on the client side."
+    guard compression.supported else {
+      return .failure(.unsupportedMessageEncoding)
+    }
+
+    let reader = LengthPrefixedMessageReader(
+      mode: .client,
+      compressionMechanism: compression,
+      logger: logger
+    )
+
+    return .success(.init(expectedCount: responseArity, reader: reader))
+  }
+
+  /// Parses the response trailers ("Trailers" in the specification) from the server into
+  /// a `GRPCStatus`.
+  ///
+  /// See: https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#responses
+  ///
+  /// - Parameter trailers: Trailers to parse.
+  private func parseTrailers(_ trailers: HTTPHeaders) -> GRPCStatus {
+    // Extract the "Status"
+    let code = trailers[GRPCHeaderName.statusCode].first
+      .flatMap(Int.init)
+      .flatMap(GRPCStatus.Code.init) ?? .unknown
+
+    // Extract and unmarshall the "Status-Message"
+    let message = trailers[GRPCHeaderName.statusMessage].first
+      .map(GRPCStatusMessageMarshaller.unmarshall)
+
+    return .init(code: code, message: message)
```

Comment:
Wouldn't we also need to extract other (custom) trailing metadata in case the server sent it, or is that done elsewhere?

Reply:
At the moment (here and in the current implementation) we don't remove the gRPC status code or message from the headers and I don't think there's much value in doing so (it would also require a copy of the headers). We can just forward on the same headers we feed into the state machine to extract the status.

Reply:
Sorry, what I meant is, where is the trailing metadata exposed to the client? Lower down the chain? I.e. we can just ignore it here, but it does get used elsewhere? (This is just for understanding, the current approach is most likely the right one.)

Reply:
The state machine will be used within a channel handler. When it reads the trailers they'll be passed into the state machine for validation, if they're successfully validated then we get a `GRPCStatus` in response which will be forwarded along with the trailers to the next handler. If they're not valid then an error will be fired and the next handler will have to synthesise a status so the handler can fail the trailing metadata promise and succeed the status promise.

Reply:
Thank you for the explanation! I do wonder whether `receiveEndOfResponseStream` should return both `GRPCStatus` and the trailers to the channel handler calling it (feels like it would be the responsibility of the state machine to extract the trailers, not of the channel handler), but I'm happy to suspend this discussion once the channel handler is actually written.

Reply:
No problem. 

> I'm happy to suspend this discussion once the channel handler is actually written.

Makes sense!